### PR TITLE
elb_application_lb: fix issue with botocore parameter validation on Rules. Fixes #27332

### LINKED ISSUE #27332: elb_application_lb module: ParamValidationError in botocore when specifying "Rules" on a new application load balancer
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

elb_application_lb

##### ANSIBLE VERSION

```
ansible 2.4.0 (devel 498bf4b5be) last updated 2017/07/26 15:27:07 (GMT +000)
  config file = None
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /ansible/.ansible-fork/lib/ansible
  executable location = /ansible/.ansible-fork/bin/ansible
  python version = 2.7.12 (default, Nov 19 2016, 06:48:10) [GCC 5.4.0 20160609]
```

##### CONFIGURATION

Default configuration

##### OS / ENVIRONMENT

ubuntu 16.04

##### SUMMARY

Parameter validation error in `boto` when running `elb_application_lb` task requiring updates to listeners.

##### STEPS TO REPRODUCE

Run a task definition like the following on to create an application load balancer:

```yaml
elb_application_lb:
  name: my-application-lb
  region: eu-west-1
  scheme: internet-facing
  subnets:
  - my-subnet
  security_groups:
  - my-sg
  listeners:
  - Protocol: HTTP
    Port: 80
    DefaultActions:
    - Type: forward
      TargetGroupName: my-target-group
    Rules:
    - Conditions:
      - Field: host-header
        Values:
        - www.example.com
      Priority: '1'
      Actions:
      - TargetGroupName: my-target-group
        Type: forward
  state: present
```

The following error is produced in boto:

```
Unknown parameter in input: "Rules", must be one of: LoadBalancerArn, Protocol, Port, SslPolicy, Certificates, DefaultActions
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_LtFMN2/ansible_module_elb_application_lb.py\", line 1011, in <module>\n    main()\n  File \"/tmp/ans
ible_LtFMN2/ansible_module_elb_application_lb.py\", line 1006, in main\n    create_or_update_elb(connection, connection_ec2, module)\n  File \"/tmp/ansible_LtFMN2/ansible_module_elb_application_lb.py\", line 901, in create_or_update_elb\n
listener_changed = create_or_update_elb_listeners(connection, module, elb)\n  File \"/tmp/ansible_LtFMN2/ansible_module_elb_application_lb.py\", line 722, in create_or_update_elb_listeners\n    connection.create_listener(**listener_to_add)\n
 File \"/usr/local/lib/python2.7/dist-packages/botocore/client.py\", line 310, in _api_call\n    return self._make_api_call(operation_name, kwargs)\n  File \"/usr/local/lib/python2.7/dist-packages/botocore/client.py\", line 573, in _make_api_
call\n    api_params, operation_model, context=request_context)\n  File \"/usr/local/lib/python2.7/dist-packages/botocore/client.py\", line 628, in _convert_to_request_dict\n    api_params, operation_model)\n  File \"/usr/local/lib/python2.7/
dist-packages/botocore/validate.py\", line 291, in serialize_to_request\n    raise ParamValidationError(report=report.generate_report())\nbotocore.exceptions.ParamValidationError: Parameter validation failed:\nUnknown parameter in input: \"Ru
les\", must be one of: LoadBalancerArn, Protocol, Port, SslPolicy, Certificates, DefaultActions\n", "module_stdout": "", "msg": "MODULE FAILURE", "rc": 0}
```

The fix is straightforward enough, and I'll submit a PR with a test.

```yaml
elb_application_lb:
  name: my-application-lb
  region: eu-west-1
  scheme: internet-facing
  subnets:
  - my-subnet
  security_groups:
  - my-sg
  listeners:
  - Protocol: HTTP
    Port: 80
    DefaultActions:
    - Type: forward
      TargetGroupName: my-target-group
    Rules:
    - Conditions:
      - Field: host-header
        Values:
        - www.example.com
      Priority: '1'
      Actions:
      - TargetGroupName: my-target-group
        Type: forward
  state: present
```

##### EXPECTED RESULTS

No error from boto

##### ACTUAL RESULTS

Error displayed in boto

```
ansible-playbook playbook.yml
```

### PR DESCRIPTION
**Pull request by mtinning**
##### SUMMARY

Fix to issue in `lib\ansible\modules\cloud\amazon\elb_application_lb.py` that was causing `ParameterValidationException` in `boto`.

Removing invalid parameters from listener args before calling `boto`'s `connection.create_listener` or `connection.modify_listener`. Ensuring easy updating by checking against list of known valid values.

Added tests.

Fixes #27332 

##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT NAME

`elb_application_lb`

##### ANSIBLE VERSION

```
ansible 2.4.0 (issue-27332-elb-application-lb-param-validation 476893ca8b) last updated 2017/07/26 17:05:18 (GMT +000)
  config file = None
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /ansible/.ansible-fork/lib/ansible
  executable location = /ansible/.ansible-fork/bin/ansible
  python version = 2.7.12 (default, Nov 19 2016, 06:48:10) [GCC 5.4.0 20160609]
```

##### ADDITIONAL INFORMATION

See #27332

### CODE REVIEW

File: lib/ansible/modules/cloud/amazon/elb_application_lb.py [hunk 1]
```diff
@@ -399,6 +399,12 @@ def _get_subnet_ids_from_subnet_list(subnet_list):
     return subnet_id_list
 
 
+def get_valid_listener_parameters(listener):
+    # See CreateListenerInput in https://github.com/boto/botocore/blob/f51271ac3ce4947ab3bbe46b82bdd79d8be0d5c2/botocore/data/elbv2/2015-12-01/service-2.json
+    valid_parameters = ['LoadBalancerArn', 'Protocol', 'Port', 'DefaultActions', 'SslPolicy', 'Certificates']
+    return {key: value for key, value in listener.items() if key in valid_parameters}
```
Comment (abadger) [inline lib/ansible/modules/cloud/amazon/elb_application_lb.py]: Just a driveby comment: for pyhton-2.6 compatibility we cannot use dict comprehensions.  However, this syntax is almost equivalent:
```
return dict((key, value) for key, value in listener.items() if key in valid_parameters)
```

**Comment by wimnat** [comment]
This just negates user error.  The reason the error is thrown is because the user entered an InvalidParameter (Rules).  By filtering out all invalid parameters with this function you could be hiding relevant errors from users.  I would just change this to remove the `Rules` key before update is called.

**Comment by mtinning** [comment]
@wimnat `Rules` is listed as a valid parameter in one of the examples (see [Line 177](https://github.com/mtinning/ansible/blob/dc3e98ff39cdb9361f68037acf08b302bc538b5b/lib/ansible/modules/cloud/amazon/elb_application_lb.py#L177))

In addition, in order to add rules to the ELB `'Rules'` is expected to be present in the parameters the user defines (see [Line 753](https://github.com/mtinning/ansible/blob/dc3e98ff39cdb9361f68037acf08b302bc538b5b/lib/ansible/modules/cloud/amazon/elb_application_lb.py#L753), amongst others).

I agree that `'Rules'` is an invalid parameter to pass to boto, however given the current design I disagree that this is user error as the module specification is being followed.

What do you mean by:

> I would just change this to remove the `Rules` key before update is called.

Which "update" are you referring to?

**Comment by s-hertel** [comment]
Hi @mtinning. The issue with this is that listeners won't be updated with rules upon creation. You'd have to run the same playbook twice to get the desired result. The parameter is valid, just not for creation, so it needs to be temporarily removed. Then the created listener needs to be added to the listeners to be updated. I've made an alternative fix here: #27379

- edit
...and actually, since you've added that parameter validation to modify listeners, I don't think the listener will get the appropriate rules even after rerunning the playbook.

**Comment by mtinning** [comment]
@s-hertel thanks for your comment.

The `Rules` parameter is valid for neither creation nor modification of the ELB - see the [parameters listed in the docs](http://boto3.readthedocs.io/en/latest/reference/services/elbv2.html#ElasticLoadBalancingv2.Client.modify_listener). As a result I don't think your #27379 fixes the issue completely.

> You'd have to run the same playbook twice to get the desired result.

I don't agree with this. There is an issue that after creating the listener the `ListenerArn` has not been set - I made a change locally to get this from the value returned from `create_listener`, which I will put up for PR once this has been approved.

**Comment by s-hertel** [comment]
@mtinning Thanks for the reply.

Sorry, I could have chosen my wording more carefully. I did not mean boto3's modify_listener() function, but that modification to the listeners happens after verifying that the desired ones are present. Here are the docs responsible for adding the rules:
http://boto3.readthedocs.io/en/latest/reference/services/elbv2.html#ElasticLoadBalancingv2.Client.create_rule
Here's where the rules are used:
https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/elb_application_lb.py#L743-L778

Since it requires the listener's ARN it cannot be done during creation. It would be awesome if you could test my PR to see if it solves your issue correctly, because it may be wrong.

> I made a change locally to get this from the value returned from create_listener, which I will put up for PR once this has been approved.

That fix is also in my PR.

**Comment by s-hertel** [comment]
@mtinning I checked out your change again with more of an open mind (I disregarded it when I was met with an ugly traceback, I'm afraid), and the way you're approaching this seems totally fine. Once you add the created listener to the listeners list for future modification I'll review this. Also, for your tests you could use the placebo fixtures. Not sure how helpful it is just to test that the functions are being called with appropriate arguments, but it would be great to have some tests (integration tests also an option).

**Comment by s-hertel** [comment]
(another musing - you'll also need to update the current_listeners in addition to adding the created listener to listeners.)

**Comment by wimnat** [comment]
I still don't think this is an appropriate fix.  If i add in some random parameter to my listener then i'll never be notified i've done something wrong.

E.g.

`
elb_application_lb:
  listeners:
    stawberries: "bla"
    ...
  ...
`

The above task would run without error.  I think it's important to tell the user that `stawberries` is an invalid parameter.

As I said, i think a better approach is just to pop the Rules before you call create or modify.

> I don't agree with this. There is an issue that after creating the listener the ListenerArn has not been set - I made a change locally to get this from the value returned from create_listener, which I will put up for PR once this has been approved.

@mtinning - Can you expand on why this is necessary?  Before creating rules i call out to get the ListenerArn - https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/elb_application_lb.py#L751

**Comment by mtinning** [comment]
@wimnat no dispute from me on the merit of `pop`ing `Rules` instead of removing all invalid parameters.

> I don't agree with this. There is an issue that after creating the listener the ListenerArn has not been set - I made a change locally to get this from the value returned from create_listener, which I will put up for PR once this has been approved.

The `ListenerArn` is needed before rule creation, in `compare_rules` - https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/elb_application_lb.py#L675. This is the line that is causing the issue.

I'll make the requested change.

**Comment by mtinning** [comment]
@s-hertel thanks for taking another look. I appreciate an initially failing build producing an "ugly traceback" makes for an unappealing PR, however if the barrier to entry was ensuring a fully working build before making a PR I wouldn't be in a position to contribute. I'll make the `ListenerArn` update.

**Comment by mtinning** [comment]
I've now added the fix for the missing `ListenerArn` issue (the new listener has been added to `current_listeners`.

@s-hertel I spent some time looking at the `placebo` module and the `placeboify` fixture. They look excellent, and I agree they would make for much better tests. However I can't currently spare the time to set them up so I have stuck with the current approach. There are obvious drawbacks, as the mocks take a bit of effort to set up and the tests are currently quite strongly coupled to the internal structure of the module, but I think some tests is better than no tests. If you disagree I can remove them.

@wimnat I changed the implementation to just `pop` the `Rules` parameter.

Thanks for all your comments.

**Comment by wimnat** [comment]
Thanks @mtinning. Just fix up ansible test sanity errors and it LGTM. shipit

**Comment by mtinning** [comment]
@wimnat thanks - sanity tests now passing (unless I've missed something?)

**Comment by wimnat** [comment]
LGTM

File: lib/ansible/modules/cloud/amazon/elb_application_lb.py [hunk 2]
```diff
@@ -719,14 +719,20 @@ def create_or_update_elb_listeners(connection, module, elb):
     for listener_to_add in listeners_to_add:
         try:
             listener_to_add['LoadBalancerArn'] = elb['LoadBalancerArn']
-            connection.create_listener(**listener_to_add)
+            # Rules is not a valid parameter for create_listener
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/elb_application_lb.py:722]: Looks good! On line [744](https://github.com/ansible/ansible/pull/27333/files#diff-a1bcebad578f0e6bfd841049d327a587L744) can you please also fix that to be deepcopy(listeners) (so the invocation gets shown correctly)? 
And lines [691](https://github.com/ansible/ansible/pull/27333/files#diff-a1bcebad578f0e6bfd841049d327a587L691) and 692 need to be set equal for new_rule, not current_rule. After that is appears rules are being updated correctly for me. Thanks for working on this.

**Review by s-hertel — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/amazon/elb_application_lb.py (hunk 2 above)
Comment (mtinning) [reply lib/ansible/modules/cloud/amazon/elb_application_lb.py:722]: Good spot for lines 691 & 692 - I'll fix this.

Hope I'm not being dim, but what does "so the invocation gets shown correctly" refer to? The change you suggest for line 744 seems harmless but I don't understand the reason for it at the moment.

Thanks!

**Comment by mtinning** [comment]
@s-hertel pushed up the requested changes so that I can knock this off before the weekend.

Thanks!

File: lib/ansible/modules/cloud/amazon/elb_application_lb.py (hunk 2 above)
Comment (s-hertel) [reply lib/ansible/modules/cloud/amazon/elb_application_lb.py:722]: When a task is run with extra verbosity it should display the arg spec the module used after the 'invocation' key (run playbook with -vvv). I could see that somewhere that actual dictionary was being modified, rather than a copy of it. 
I thought I kept running the wrong playbook or was doing something wrong syntactically because it was deleting my rules:
```
...
            "listeners": [
                {
                    "DefaultActions": [
                        {
                            "TargetGroupArn": "arn:aws:elasticloadbalancing:us-east-1:448830907657:targetgroup/sherteltg/bd83c5c780de26d7",
                            "Type": "forward"
                        }
                    ],
                    "ListenerArn": "arn:aws:elasticloadbalancing:us-east-1:448830907657:listener/app/shertel-application-lb/365334adfd1f3606/9118ad2fccc38cde",
                    "Port": 80,
                    "Protocol": "HTTP",
                    "Rules": []
                }
            ],
```

**Review by s-hertel — APPROVED** [review]

File: test/units/modules/cloud/amazon/test_elb_application_lb.py [hunk 3]
```diff
@@ -0,0 +1,166 @@
+#
+# (c) 2017 Michael Tinning
```
Comment (ryansb) [inline test/units/modules/cloud/amazon/test_elb_application_lb.py:2]: Copyright guidelines have changed recently to be much shorter (and easier to follow) [here](http://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#copyright)

So you can delete the whole license snippet below this and add:

```
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
```

File: test/units/modules/cloud/amazon/test_elb_application_lb.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+from nose.plugins.skip import SkipTest
+import json
+import pytest
+from copy import deepcopy
+from ansible.module_utils._text import to_bytes
+from ansible.module_utils import basic
+from ansible.module_utils.ec2 import HAS_BOTO3
+
+if not HAS_BOTO3:
+    raise SkipTest("test_api_gateway.py requires the `boto3` and `botocore` modules")
```
Comment (ryansb) [inline test/units/modules/cloud/amazon/test_elb_application_lb.py]: s/test_api_gateway.py/test_elb_application_lb.py/

File: test/units/modules/cloud/amazon/test_elb_application_lb.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+        LoadBalancerArn=existing_elb['LoadBalancerArn']
+    )
+
+
+def test_compare_rules_called_with_new_listener(
+    mocker,
+    connection,
+    listener,
+    elb,
+    compare_listeners,
```
Comment (ryansb) [inline test/units/modules/cloud/amazon/test_elb_application_lb.py:139]: You don't need to re-include some of these fixtures, as they are used in the `elb` fixture but never called in this test.

**Review by ryansb — CHANGES_REQUESTED** [review]

File: test/units/modules/cloud/amazon/test_elb_application_lb.py (hunk 5 above)
Comment (mtinning) [reply test/units/modules/cloud/amazon/test_elb_application_lb.py:139]: All of these fixtures are called in this test (believe it or not!)

**Comment by mtinning** [comment]
@ryansb thanks for your comments - I've made the requested changes

**Comment by mtinning** [comment]
@ryansb is there any chance we can try to get this in? Is there anything else you'd like to request? I'd like to be working off the main ansible devel branch as fixes are rolling in that I need, but this issue prevents that. Many thanks.

**Review by ryansb — APPROVED** [review]

**Comment by ryansb** [comment]
rebuild_merge
